### PR TITLE
feature: add more signals - language signal

### PR DESCRIPTION
**Pull request by liavweiss**
# Add Language Signal Types
Description - https://github.com/vllm-project/semantic-router/pull/1052#issuecomment-3746528584

### CODE REVIEW

**Comment by liavweiss** [comment]
This PR implements **2 new signal types**:
- **Complexity Signal**: Language-agnostic query complexity classification (simple/medium/complex)
- **Language Signal**: Multi-language detection using `whatlanggo` library (100+ languages)

Both signals are fully integrated with parallel execution, 28 unit tests, and documentation updates.
I'm open to implementing **additional signal types** if needed (e.g., user tier signals, context length optimization, or other signals based on your routing requirements). Happy to discuss what would be most valuable!

**Comment by Xunzhuo** [comment]
thanks! this is wonderful!

**Review by Xunzhuo — APPROVED** [review]
some followups:
1. investigate more advancing algorithm to determine if a prompt is complex or simple (maybe based on ML techiniques)
2. update signals page in dashboard UI to virtualize the signals

File: src/semantic-router/pkg/classification/complexity_classifier.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+	logging.Infof("Complexity classification: type=%s, score=%.2f, confidence=%.2f", complexityType, score, confidence)
+
+	return &ComplexityResult{
+		ComplexityType: complexityType,
+		Confidence:     confidence,
+	}, nil
+}
+
+// calculateComplexityScore calculates a complexity score (0.0-1.0) based on language-agnostic heuristics
```
Comment (Xunzhuo) [inline src/semantic-router/pkg/classification/complexity_classifier.go]: some very short prompts are the too complex task as well, so let us dig deeper in this area

**Comment by rootfs** [comment]
@liavweiss can you break this down into 2, one for language detector (which is easy to understand) and complexity (this one needs more thinking and validation, it is not clear the detector can match up with the complexity views of model evaluations). The first one can be merged quickly, the 2nd one can move on as a research initiative.

**Comment by rootfs** [comment]
@liavweiss @Xunzhuo  a side note, would it make sense to add 1) a latency signal based on TPOT? That'll be a step to latency sensitive scenarios, and 2) a pricing signal that use token cost to drive model selection

**Comment by liavweiss** [comment]
As we discussed above @Xunzhuo , @rootfs - I separated and kept this PR with the language signal only.
I will open a separate research issue for the complexity signal.


## What's Included

### ✅ Language Signal Implementation

- **Language Classifier**: Uses `whatlanggo` library for robust language detection (100+ languages supported)
- **ISO 639-1 Codes**: Returns standard language codes (e.g., "en", "es", "zh", "fr", "de", "ja")
- **Confidence Scores**: Provides reliability metrics for language detection
- **Fallback Handling**: Defaults to "en" if detection is unreliable

### ✅ Integration

- **Signal Evaluation**: Integrated into parallel signal evaluation pipeline
- **Decision Engine**: Language signals can be used in decision rules with AND/OR operators
- **Configuration**: Supports both legacy and Python CLI config formats
- **Dashboard UI**: Language signal visible in dashboard (Python CLI format)

### ✅ Testing

- **16 unit tests** covering:
  - Multiple languages (English, Spanish, Russian, Chinese, French, German, Japanese)
  - Edge cases (empty text, very long strings, mixed languages, special characters, unicode, code snippets)
  - Confidence scoring
- **All tests passing**: 149/149 unit tests pass

### ✅ Documentation

- Updated signal count from 6 to 7 types
- Added Language Signal section to documentation
- Updated configuration examples
- Updated Mermaid diagrams

## Configuration Example

```yaml
signals:
  language:
    - name: "en"
      description: "English language queries"
    - name: "es"
      description: "Spanish language queries"
    - name: "zh"
      description: "Chinese language queries"
    - name: "ru"
      description: "Russian language queries"
    - name: "fr"
      description: "French language queries"
    - name: "de"
      description: "German language queries"
    - name: "ja"
      description: "Japanese language queries"

decisions:
  - name: "spanish_route"
    priority: 100
    rules:
      operator: "AND"
      conditions:
        - type: "language"
          name: "es"
    modelRefs:
      - model: "spanish-model"
```

## Use Cases

- Route queries to language-specific models
- Apply language-specific policies (e.g., PII handling, system prompts)
- Support multilingual applications
- Enable language-aware caching and optimization

## Technical Details

- **Library**: `whatlanggo` (Go library for language detection)
- **Latency**: <1ms
- **Supported Languages**: 100+ languages via ISO 639-1 codes
- **Reliability**: Automatic fallback to English for unreliable detections

## Testing

All unit tests pass. Language detection verified for:
- English, Spanish, Russian, Chinese, French, German, Japanese
- Edge cases: empty text, mixed languages, unicode, code snippets

## Related

- Issue: #983
